src/storage/src/table/mod.rs

```diff
@@ -130,28 +130,22 @@ where
 /// Collects data chunks from stream of rows.
 pub async fn collect_data_chunk_with_builder<E, S>(
     stream: &mut S,
-    chunk_size: Option<usize>,
     builder: &mut DataChunkBuilder,
 ) -> Result<Option<DataChunk>, E>
 where
     S: Stream<Item = Result<OwnedRow, E>> + Unpin,
 {
-    for _ in 0..chunk_size.unwrap_or(usize::MAX) {
-        match stream.next().await.transpose()? {
-            Some(row) => {
-                builder.append_one_row_no_finish(row);
-            }
-            None => break,
+    // TODO(kwannoel): If necessary, we can optimize it in the future.
+    // This can be done by moving the check if builder is full from `append_one_row` to here,
+    while let Some(row) = stream.next().await.transpose()? {
+        let result = builder.append_one_row(row);
+        if let Some(chunk) = result {
+            return Ok(Some(chunk));
         }
     }
 
-    let chunk = builder.build_data_chunk();
-
-    if chunk.cardinality() == 0 {
-        Ok(None)
-    } else {
-        Ok(Some(chunk))
-    }
+    let chunk = builder.consume_all();
+    Ok(chunk)
```

Comment:
This seems a little bit weird to me. I assume this method will be called in a loop. If we always consume the chunk builder here, then what's the difference compared to initializing a `ChunkBuilder` inside this function?

Reply:
Actually it was also unintuitive to me. The reference scenario is here: https://github.com/risingwavelabs/risingwave/issues/11007#issuecomment-1681914555.

But idea is this. Suppose snapshot read proceeds midway.

It can get stuck here while waiting for next row, perhaps object store latency:
```
            Some(row) => {
    while let Some(row) = stream.next().await.transpose()? 
...
```

Now there are some rows in the buffer before this, but they are not yet flushed.

If some upstream updates come, we will process those, and if within the upstream updates, there is a barrier, we will drop the snapshot.

But critically even though the snapshot is dropped, there are still rows inside the buffer.

We don't want to just drop these rows, losing the progress we made.

So we should still flush these rows out, and that is done outside of the snapshot_read function, only on barrier.

Reply:
Yeah, this makes much sense. In other words, this stream will likely be canceled in its usage cases.